lib/bolt/plugin/puppetdb.rb

```diff
@@ -16,8 +16,35 @@ def hooks
       end
 
       def lookup_targets(opts)
-        nodes = @puppetdb_client.query_certnames(opts['query'])
-        nodes.map { |certname| { 'uri' => certname } }
+        targets = @puppetdb_client.query_certnames(opts['query'])
+        facts = []
+        # This is done in parts instead of in place so that we only need to
+        # make one puppetDB query
+        Bolt::Util.walk_vals(opts['config']) do |value|
+          if value.is_a?(String)
+            facts << value.split(".")
+          end
+          value
+        end
+        fact_values = @puppetdb_client.fact_values(targets, facts)
+
+        targets.map do |certname|
+          target = { 'uri' => certname }
+          # Produces [{certname => 'foo', 'value' => 'bar'}, {...}]
+          target_data = fact_values.select { |data| data['certname'] == certname }
+          target['config'] = resolve_config(opts['config'], target_data)
+          target
+        end.compact
+      end
+
+      def resolve_config(config, target_data)
+        Bolt::Util.walk_vals(config) do |value|
```

Comment:
Is this too expensive to do twice? I'm not sure if there's a clever way to avoid it.